src/BookProvider.php

```diff
@@ -321,66 +326,46 @@ protected function startCredits( Book $book, array $chapters, array $otherPages,
 		if ( $book->scan != '' ) {
 			$pages[] = 'Index:' . $book->scan;
 		}
+
 		$pages = array_unique( array_merge( $pages, $otherPages ) );
-		foreach ( $this->splitArrayByLength( $pages, 3000 ) as $batch ) {
-			$params = [
-				'lang' => $book->lang, 'format' => 'json', 'page' => implode( '|', $batch )
-			];
-			$promises[] = $this->api->getAsync(
-				$this->creditUrl,
-				[ 'query' => $params ]
-			);
-		}
+		$pageCredits = $this->creditRepo->getPageCredits( $book->lang, $namespaces, $pages );
 
 		$imagesSet = [];
 		foreach ( $pictures as $id => $picture ) {
 			if ( $picture->name ) {
 				$imagesSet[$picture->name] = true;
 			}
 		}
+		$imageCredits = [];
 		if ( !empty( $imagesSet ) ) {
 			$images = array_keys( $imagesSet );
-			foreach ( $this->splitArrayByLength( $images, 3000 ) as $batch ) {
-				$params = [
-					'lang' => $book->lang,
-					'format' => 'json',
-					'image' => implode( '|', $batch ),
-				];
-				$promises[] = $this->api->getAsync( $this->creditUrl, [ 'query' => $params ] );
-			}
+			$imageCredits = $this->creditRepo->getImageCredits( $book->lang, $images );
 		}
 
-		return $promises;
+		return array_merge( $pageCredits, $imageCredits );
 	}
 
 	/**
-	 * @param PromiseInterface[] $promises
+	 * @param array $creditResults
 	 * @return array
 	 */
-	public function finishCredit( $promises ) {
+	public function finishCredit( $creditResults ) {
```

Comment:
Great point! I removed this `finishCredit` method but I'm afraid the results still need to be ordered since the image credits and page credits might return the same `actor names` so we need to combine and count their contributions? Does this make sense?

Reply:
Oh yes, I totally wasn't thinking correctly. Good point. I wonder if the two queries could be done in one request, with a UNION. Then the sorting can be done by the DB. (Not sure it matters too much though.)

Reply:
I left the queries separately  so we don't have a huge query. But do think is better to make one request? Do you think that might be a better approach for performance maybe?

Reply:
You can't! The image query is on commonswiki (s7) while the wikisource databases live on a different db section. In my testing both queries are pretty fast (at least on the new cluster), so combining them even if we could probably wouldn't see significant gains.